cpp/arcticdb/storage/s3/tcp_ping_ec2.hpp

```diff
@@ -0,0 +1,68 @@
+#include <folly/net/NetOps.h>
+#include <arcticdb/log/log.hpp>
+
+namespace arcticdb::storage::s3 {
+
+namespace {
+using namespace folly::netops;
+
+struct RaiiSocket {
+    static constexpr folly::NetworkSocket invalid{};
+    folly::NetworkSocket fd;
+
+    RaiiSocket() : fd(folly::netops::socket(AF_INET, SOCK_STREAM, 0)) {}
+
+    template<typename Fun>
+    int attempt([[maybe_unused]] const char* step, Fun&& fun) {
+        if (fd != invalid) {
+            int out = fun(fd);
+            if (out < 0) {
+                close(fd);
+                fd = invalid;
+                ARCTICDB_DEBUG(arcticdb::log::Loggers::instance()->storage(),
+                    "ec2_metadata step {} failed with {}. errno={}", step, out, errno);
+            }
+            return out;
+        }
+        return -1;
+    }
+
+    ~RaiiSocket() {
+        if (fd != invalid) {
+            close(fd);
+        }
+    }
+};
+}
+
+/* Quick check without incurring the default, hard-coded long wait time in the AWS SDK. */
+bool ec2_metadata_endpoint_reachable() {
+    using namespace folly::netops;
+
+    RaiiSocket s;
+    s.attempt("set_socket_non_blocking", &set_socket_non_blocking);
+
+    s.attempt("connect", [](auto& fd) {
+        sockaddr_in addr;
+        addr.sin_family = AF_INET;
+        addr.sin_port = htons(80);
+        addr.sin_addr.s_addr = inet_addr("169.254.169.254");

```

Comment:
What is this IP?

Reply:
This is the well-known AWS metadata endpoint IP. You can google it for more details.

Reply:
Should have a comment really, please let's not assume that everyone working on this project has a list of memorized IP addresses!

Reply:
I think the name of the function makes it clear what this IP is?